react-native-community/react-native-cameraroll
# Added `include` parameter to getPhotos to let users tradeoff performance by omitting metadata

# Summary

On a CameraRoll with ~30k photos, `getPhotos` was taking >3 minutes to get the metadata for all the photos. This was too slow for our application.

Digging through the ObjectiveC code, I saw that it was fetching detailed metadata for every photo, which took a lot of time. The metadata was only used to populate the filename and to filter by mimetype.

## Fixes

### Adding `include` to `getPhotos`

1. Adds an `include` param to `getPhotos`. Users can include `fileName`, `fileSize`, or `location`. The fields are null unless explicitly included
2. On iOS: Avoid calling `[PHAssetResource assetResourcesForAsset:asset]` in cases where it's not needed to dramatically improve performance
   - Most of the performance gain is here: this reduced run-times from 6 to 0.6 seconds
3. On iOS: Set `assetFetchOptions.fetchLimit = first;` when we know we won't be omitting any returned items
   - This only reduced running time by another 60%, down to ~0.2 seconds
4. On Android: Avoid calling `putLocationInfo` if the location is not needed, which reduces running time by 30%

With this fix, it takes 0.2 instead of 6 seconds to fetch 1k photos on iOS

### Making `toDate` consistently inclusive

- Made the logic on iOS for comparing `toDate` the same as that on Android by using `<=` instead of `<`

### Enabled Prettier for Typescript

- Changed the config to make it work for Typescript. It still isn't used for README.md, but a future quick commit can fix that

## Test Plan

Added automated test to verify the native module is tested correctly.

Also created an example app screen for testing.

1. Run the example app with `yarn start:ios`
2. Switch to the GetPhotosPerformanceExample.
3. Try tapping on both the buttons and see the speed difference. Ensure you have at least some photos on the phone. Here's a test with ~500 photos dragged onto it

| iOS | Android |
| --- | --- |
| ![iOS](https://user-images.githubusercontent.com/2937410/82715579-a0f1c900-9c48-11ea-874a-8ddfef53e023.gif) | ![Android](https://user-images.githubusercontent.com/2937410/82715586-af3fe500-9c48-11ea-8da7-33c2ecd4afb1.gif) |

### What's required for testing (prerequisites)?

Just `yarn install`, then `yarn start:ios` and `yarn start:android`

### What are the steps to reproduce (after prerequisites)?

## Compatibility

| OS      | Implemented |
| ------- | :---------: |
| iOS     |    ✅     |
| Android |    ✅ (unchanged behavior, but API is there)     |

## Checklist
